module/x/peggy/keeper/pool.go

```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		return 0, err
+	}
+
+	// add a second index with the fee
+	idxKey := types.GetFeeSecondIndexKey(fee)
+	var idSet types.IDSet
+	if store.Has(idxKey) {
+		bz := store.Get(idxKey)
+		k.cdc.UnmarshalBinaryBare(bz, &idSet)
+	}
+	idSet = append(idSet, nextID)
+	store.Set(idxKey, k.cdc.MustMarshalBinaryBare(idSet))
+	return nextID, nil
+}
+
+func (k Keeper) setPoolEntry(ctx sdk.Context, id uint64, val types.OutgoingTx) error {
+	bz, err := k.cdc.MarshalBinaryBare(val)
+	if err != nil {
+		return err
+	}
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.GetOutgoingTxPoolKey(id), bz)
+	return nil
+}
+
+func (k Keeper) getPoolEntry(ctx sdk.Context, id uint64) (*types.OutgoingTx, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetOutgoingTxPoolKey(id))
+	if bz == nil {
+		return nil, types.ErrUnknown
+	}
+	var r types.OutgoingTx
+	k.cdc.UnmarshalBinaryBare(bz, &r)
+	return &r, nil
+}
+
+func (k Keeper) GetCounterpartDenominator(ctx sdk.Context, voucherDenom string) (*types.BridgedDenominator, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetDenominatorKey(voucherDenom))
+	if bz == nil {
+		return nil, sdkerrors.Wrap(types.ErrUnknown, "denominator")
+	}
+	var bridgedDenominator types.BridgedDenominator
+	return &bridgedDenominator, k.cdc.UnmarshalBinaryBare(bz, &bridgedDenominator)
+}
+
+func (k Keeper) SetCounterpartDenominator(ctx sdk.Context, bridgeContractAddr, tokenID string) {
+	store := ctx.KVStore(k.storeKey)
+	voucherDenominator := toVoucherDenominator(bridgeContractAddr, tokenID)
+	bridgedDenominator := types.BridgedDenominator{
+		BridgeContractAddress: bridgeContractAddr,
+		TokenID:               tokenID,
+	}
+	store.Set(types.GetDenominatorKey(voucherDenominator[voucherPrefixLen:]), k.cdc.MustMarshalBinaryBare(bridgedDenominator))
+}
+
+func (k Keeper) HasCounterpartDenominator(ctx sdk.Context, voucherDenominator string) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Has(types.GetDenominatorKey(voucherDenominator))
+}
+
+func (k Keeper) IterateOutgoingPoolByFee(ctx sdk.Context, cb func(uint64, types.OutgoingTx) bool) error {
+	prefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.SecondIndexOutgoingTXFeeKey)
+	iter := prefixStore.ReverseIterator(nil, nil)
+	for ; iter.Valid(); iter.Next() {
+		var ids types.IDSet
+		k.cdc.MustUnmarshalBinaryBare(iter.Value(), &ids)
+		// cb returns true to stop early
+		for _, id := range ids {
+			tx, err := k.getPoolEntry(ctx, id)
+			if err != nil {
+				return err
+			}
+			if cb(id, *tx) {
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+func (k Keeper) autoIncrementID(ctx sdk.Context, idKey []byte) uint64 {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(idKey)
+	var id uint64 = 1
+	if bz != nil {
+		id = binary.BigEndian.Uint64(bz)
+	}
+	bz = sdk.Uint64ToBigEndian(id + 1)
+	store.Set(idKey, bz)
+	return id
+}
+
+func toVoucherDenominator(contractAddr, token string) string {
```

Comment:
could you add more more comments here about what exactly your doing to make the voucher denominator? Just for ease of understanding.